packages/api/src/controllers/stream.ts

```diff
@@ -956,60 +921,99 @@ app.put(
-      ) {
-        // last session is too old, probably transcoding wasn't happening, and so there
-        // will be no recording
-      } else {
-        const userSessionId = session.previousSessions?.[0] ?? session.id;
-        const userSession = await db.session.get(userSessionId);
-        await queue.delayedPublishWebhook(
-          "events.recording.ready",
-          {
-            type: "webhook_event",
-            id: uuid(),
-            timestamp: Date.now(),
-            streamId: stream.id,
-            event: "recording.ready",
-            userId: stream.userId,
-            sessionId: session.id,
-            payload: {
-              recordingUrl: getRecordingUrl(ingest, userSession),
-              mp4Url: getRecordingUrl(ingest, userSession, true),
-              session: {
-                ...toExternalSession(userSession, ingest, true),
-                recordingStatus: "ready", // recording will be ready if this webhook is actually sent
-              },
-            },
-          },
-          USER_SESSION_TIMEOUT + HTTP_PUSH_TIMEOUT
+  await db.stream.update(stream.id, patch);
+
+  const changed =
+    stream.isActive !== patch.isActive ||
+    stream.region !== patch.region ||
+    stream.mistHost !== patch.mistHost;
+  const isStaleCleanup = !patch.isActive && isStreamStale(stream);
+
+  if (changed && !isStaleCleanup) {
+    const event = patch.isActive ? "stream.started" : "stream.idle";
+    await queue
+      .publishWebhook(`events.${event}`, {
+        type: "webhook_event",
+        id: uuid(),
+        timestamp: Date.now(),
+        streamId: stream.id,
+        event: event,
+        userId: stream.userId,
+      })
+      .catch((err) => {
+        logger.error(
+          `Error sending /setactive hooks stream_id=${stream.id} event=${event} err=`,
+          err
         );
+      });
+  }
+
+  // opportunistically trigger recording.ready logic for this stream's sessions
+  triggerSessionRecordingHooks(stream, queue, ingest).catch((err) => {
+    logger.error(
+      `Error triggering session recording hooks stream_id=${stream.id} err=`,
+      err
+    );
+  });
+}
+
+/**
+ * Trigger delayed recording.ready events for each active session in the stream.
+ * These recording.ready events aren't sent directly to the user, but instead
+ * the handler will check if the session is actually inactive to fire the hook.
+ */
+async function triggerSessionRecordingHooks(
+  streamOrSession: DBStream,
+  queue: Queue,
+  ingest: string
+) {
+  const { id, parentId } = streamOrSession;
+  let sessions = parentId
+    ? [streamOrSession]
+    : await db.stream.getActiveSessions(id);
+
+  // backward compat with child streams that didn't have isActive field
+  if (sessions.length === 0) {
+    const last = await db.stream.getLastSession(id);
+    if (last && last.isActive === undefined) {
+      sessions = [last];
+    }
+  }
+
+  for (const session of sessions) {
```

Comment:
nit: could be breakdown in two separate smaller functions for clarity
```suggestion
  for (const session of sessions) {
    try {
      await processSession(session, queue, ingest);
    } catch (err) {
      logger.error(
        `Error processing session session_id=${session.id} err=`,
        err
      );
    }
  }
}

async function processSession(
  session: DBStream,
  queue: Queue,
  ingest: string
) {
```

Reply:
As we are only talking about 7 lines of logic, I think it's kinda overkill to extract this helper, ChatGPT does some of those every now and then... I'm gonna do it anyway in case you also think it's more readable, not only copied from GPT 😄

Reply:
Done, but gave it a better name than a generic "processSession" that would only obfuscate what is done in that piece of the code. Perhaps the whole logic from `triggerSessionRecordingHooks` is more easily understandable now only be skimming through the code.